src/structured-text/structured_text_parser.cpp

```diff
@@ -0,0 +1,259 @@
+
+Module: Structured Text Language Parser
+
+Author: Matthias Weiss, matthias.weiss@diffblue.com
+
+\*******************************************************************/
+
+/// \file
+/// Structured Text Language Parser
+
+#include "structured_text_parser.h"
+#include "structured_text_parse_tree.h"
+#include <cmath>
+#include <iostream>
+#include <util/expr.h>
+
+structured_text_parsert structured_text_parser;
+
+extern char *yystructured_texttext;
+
+int yystructured_texterror(const std::string &error)
+{
+  structured_text_parser.parse_error(error, yystructured_texttext);
+  return 0;
+}
+
+void structured_text_parsert::add_function_block(exprt &block)
+{
+  INVARIANT(
+    block.id() == ID_structured_text_function_block,
+    "Root expression ID should be ID_structured_text_function_block");
+
+  // Generate new function block.
+  structured_text_parse_treet::function_blockt fb{find_name(block),
+                                                  find_version(block)};
+
+  // Fill the block with networks and variables.
+  exprt var_list = find_variable_list(block);
+  find_variables(fb, var_list);
+  exprt network_list = find_network_list(block);
+  find_networks(fb, network_list);
+
+  parse_tree.add_function_block(fb);
+}
+
+void structured_text_parsert::add_function(exprt &function)
+{
+  INVARIANT(
+    function.id() == ID_structured_text_function,
+    "Expression ID should be structured_text_function");
+
+  // Generate new function.
+  structured_text_parse_treet::functiont fn{find_name(function),
+                                            find_version(function)};
+
+  // Fill the function with networks and variables.
+  exprt var_decls = find_variable_list(function);
+  find_variables(fn, var_decls);
+  exprt network_list = find_network_list(function);
+  find_networks(fn, network_list);
+
+  parse_tree.add_function(fn);
+}
+
+void structured_text_parsert::clear()
+{
+  parsert::clear();
+  parse_tree.clear();
+}
+
+void structured_text_parsert::print_tree(std::ostream &out) const
+{
+  parse_tree.output(out);
+}
+
+irep_idt structured_text_parsert::find_name(const exprt &root)
+{
+  for(exprt op : root.operands())
+  {
+    if(op.get(ID_structured_text_type) == ID_structured_text_identifier)
+      return op.get(ID_value);
+  }
+
+  INVARIANT(false, "Root expression should have a name");
+  return irep_idt();
+}
+
+float structured_text_parsert::find_version(const exprt &root)
+{
+  for(exprt op : root.operands())
+  {
+    if(op.type().id() == ID_structured_text_version)
+      return std::stof(op.get(ID_value).c_str());
+  }
+
+  INVARIANT(false, "Root expression should have a version");
+  return -1;
+}
+
+exprt structured_text_parsert::find_variable_list(const exprt &root)
+{
+  for(exprt op : root.operands())
+  {
+    if(op.id() == ID_structured_text_var_decls)
+      return op;
+  }
+  INVARIANT(false, "Root expression should have a variable list");
+  return exprt(ID_nil);
```

Comment:
Use `nil_exprt{}` (as that uses a single, shared irept). Though you could actually also use `UNREACHABLE` and then not add any `return` statement.

Reply:
I tried that first, but Eclipse (cpplint?) complains about it. Is there any way to suppress warnings like that properly?

Reply:
Hmm, that's surprising as we already use `UNREACHABLE` in that way in several places - or maybe Eclipse also warns in those cases?